bug: improve schema checking for `insert into` cases

zhuqi-lucas opened the pull request:
## Which issue does this PR close?

Describe the bug
In, https://github.com/apache/datafusion/issues/14394, it was reported that while attempting to implement a DataSink different schemas for the record batches were being given than per the RecordBatchStream.

A fix for the given example, an INSERT INTO ... VALUES query, was merged (https://github.com/apache/datafusion/pull/14472). However, this issue likely arises when the schema of the source of an INSERT statement contain fields that differ from the table schema in terms of nullability. That is, the problem is not just limited to INSERT INTO ... VALUES statements.

- Closes [#14550](https://github.com/apache/datafusion/issues/14550)

## What changes are included in this PR?
Add a separate nullable checking besides the original checking which only include the name and datatype.
Improve the error message to including more info about the error.


We will improve the checking for the 3 cases, also improve the error message.

There are three cases we need to check
   1. The len of the schema of the plan and the schema of the table should be the same
   
2 **(The nullable flag of the schema of the plan and the schema of the table should be the same)** This is not needed, we have checking in execution plan.
   
   3. The datatype of the schema of the plan and the schema of the table should be the same

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

jayzhan211 reviewed (commented):
Can you explain the reason of the change in test.slt, thanks

jonahgao commented on datafusion/common/src/dfschema.rs:
```diff
@@ -1028,20 +1028,48 @@ impl SchemaExt for Schema {
-                f1.name() == f2.name()
-                    && DFSchema::datatype_is_logically_equal(
+            .try_for_each(|(f1, f2)| {
+                if f1.is_nullable() != f2.is_nullable() {
```
This seems a regression to me 🤔.  Even though the schema of a source is nullable, all of its data can be non-nullable, and in such cases, it can still be inserted into a non-nullable sink. When inserting, we currently validate against the actual data rather than the schema. See [check_not_null_constraints](https://github.com/apache/datafusion/blob/2c73fcd47ed7391c7fd34d6c593c6f0c455670d0/datafusion/physical-plan/src/execution_plan.rs#L1003)

jonahgao replied:
If 'DataSink receiving different schemas' is an issue, we can rewrite the schema of batches emitted by DataSinkExec.

zhuqi-lucas replied:
Thank you @jayzhan211 and @jonahgao for review, this is a good point, i change it to the only error case for nullable check:
// only check the case when the table field is not nullable and the query field is nullable

zhuqi-lucas commented on datafusion/sqllogictest/test_files/insert.slt:
```diff
@@ -78,7 +104,7 @@ physical_plan
 INSERT INTO table_without_values SELECT
 SUM(c4) OVER(PARTITION BY c1 ORDER BY c9 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING),
-COUNT(*) OVER(PARTITION BY c1 ORDER BY c9 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING)
+NULLIF(COUNT(*) OVER(PARTITION BY c1 ORDER BY c9 ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING), 0)
```
This regression now not happened after the above code changes.

zhuqi-lucas commented on datafusion/sqllogictest/test_files/aggregate_skip_partial.slt:
```diff
@@ -228,6 +231,27 @@ CREATE TABLE aggregate_test_100_null (
+statement ok
+CREATE TABLE aggregate_test_100_null (
+  c2  TINYINT NOT NULL,
+  c5  INT,
```
Noted, i also add the successful case which the c5 is nullable.

zhuqi-lucas commented:
> Can you explain the reason of the change in test.slt, thanks

Thank you for review @jayzhan211, i already updated the slt now, and added note for the only 2 different results for the sql.

jayzhan211 commented on datafusion/common/src/dfschema.rs:
```diff
@@ -1028,20 +1028,48 @@ impl SchemaExt for Schema {
+    // 1. The len of the schema of the plan and the schema of the table should be the same
+    // 2. The nullable flag of the schema of the plan and the schema of the table should be the same
+    // 3. The datatype of the schema of the plan and the schema of the table should be the same
+    fn logically_equivalent_names_and_types(&self, other: &Self) -> Result<(), String> {
```
You can also define different messages with `internal_err!("msg1"), internal_err!("msg2")`

zhuqi-lucas replied:
Thank you @jayzhan211 for this good suggestion, i change my code to use result<()>, it seems it's enough for this case, similar with many other cases, for example:

```rust
/// Check if the schema have some fields with the same name
    pub fn check_names(&self) -> Result<()> {
   }
```

zhuqi-lucas commented on datafusion/sqllogictest/test_files/insert.slt:
```diff
@@ -296,8 +296,11 @@ insert into table_without_values(field1) values(3);
-statement error Execution error: Invalid batch column at '0' has null but schema specifies non-nullable
+statement error
 insert into table_without_values(field2) values(300);
+----
```
This is due to we have checking in insert into plan now, before this we only have the check for execution plan check.

zhuqi-lucas commented:
The CI error is caused by:

https://github.com/apache/datafusion/pull/14625

zhuqi-lucas commented on datafusion/sqllogictest/test_files/insert_to_external.slt:
```diff
@@ -60,17 +60,16 @@ STORED AS parquet
 select * from dictionary_encoded_values
 ----
-2
+DataFusion error: Error during planning: Inserting query must have the same schema nullability as the table. Expected table field 'b' nullability: false, got field: 'b', nullability: true
```
This is also expected, because PARTITIONED BY (b), will make the b nullable to false.

We shouldn't support insert nullable value for partition key i think.

jayzhan211 replied:
`because PARTITIONED BY (b), will make the b nullable to false` can be add as comment

zhuqi-lucas replied:
Added comments in latest PR, thanks!

Error should be after query error to pass CI, i think it's auto generated after the PR here:

https://github.com/apache/datafusion/pull/14439

jayzhan211 replied:
It is auto generated a long time before, we need to manually move it

zhuqi-lucas replied:
@jayzhan211  This https://github.com/apache/datafusion/pull/14439/files#diff-51757b2b1d0a07b88551d88eabeba7f74e11b5217e44203ac7c6f613c0221196R273 merged less than a week ago. I think it's start from there, and i tried local sql logic test use -- --complete, it's also the same result, i guess the CI also use it to generate and verify?

jonahgao replied:
This is the multiline error feature of sqllogictest-rs.
https://github.com/risinglightdb/sqllogictest-rs/blob/7ee44cd995fb65175bb647d07f63b557dbaa22c7/CHANGELOG.md#0180---2023-11-08

jonahgao commented on datafusion/common/src/dfschema.rs:
```diff
@@ -1028,21 +1028,41 @@ impl SchemaExt for Schema {
+                .zip(other.fields().iter())
+                .try_for_each(|(f1, f2)| {
+                    // only check the case when the table field is not nullable and the insert data field is nullable
+                    if !f1.is_nullable() && f2.is_nullable() {
```
This condition would prevent the following query from executing, but it works on both the main branch and Postgres.
```sql
create table t1(a int not null);
create table t2(a int);
insert into t2 values(100);
insert into t1 select * from t2;
```
As I mentioned earlier, we already have a check during execution called [check_not_null_constraints](https://github.com/apache/datafusion/blob/2c73fcd47ed7391c7fd34d6c593c6f0c455670d0/datafusion/physical-plan/src/execution_plan.rs#L1003), so I think we should not add this restriction here.

zhuqi-lucas replied:
@jonahgao  I addressed the comments, but is the issue valid now? https://github.com/apache/datafusion/issues/14550

I can't find which bug we need to fix, thanks!

jonahgao replied:
I'm not sure if it's necessary to ensure that the schema of output batches has the same nullability. This issue exists not only with inserts but also with other queries like UNION.
```sh
DataFusion CLI v45.0.0
> create table t1(a int not null) as values(1);
0 row(s) fetched.
Elapsed 0.009 seconds.

> create table t2(a int) as values(2);
0 row(s) fetched.
Elapsed 0.011 seconds.

> select * from t1 union all select * from t2;
batch schema: Schema { fields: [Field { name: "a", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }
batch schema: Schema { fields: [Field { name: "a", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }
+---+
| a |
+---+
| 1 |
| 2 |
+---+
2 row(s) fetched.
Elapsed 0.007 seconds.
```

zhuqi-lucas replied:
Updated, it seems several union issues are also related the nullable for union:
https://github.com/apache/datafusion/issues/14352

jonahgao commented on datafusion/common/src/dfschema.rs:
```diff
@@ -1028,21 +1028,32 @@ impl SchemaExt for Schema {
+                .try_for_each(|(f1, f2)| {
+                    if f1.name() != f2.name() || !DFSchema::datatype_is_logically_equal(f1.data_type(), f2.data_type()) {
+                        _plan_err!(
+                            "Inserting query schema mismatch: Expected table field '{}' with type {:?}, \
```
Based on the function name, it seems that we haven't restricted it to only be used for insertion🤔.

zhuqi-lucas replied:
Checked the function only be called by insert into cases.

alamb replied:
Perhaps we could update the comments to reflect this

zhuqi-lucas replied:
Added comments in latest PR, thanks all.

alamb commented on datafusion/core/tests/dataframe/mod.rs:
```diff
@@ -5274,3 +5274,61 @@ async fn register_non_parquet_file() {
+    // Test case 1:
+    let write_df = session_ctx.sql("values (1, 2), (3, 4)").await.unwrap();
+
+    match write_df
+        .write_table("t", DataFrameWriteOptions::new())
+        .await
+    {
+        Ok(_) => {}
+        Err(e) => {
+            assert_contains!(
+                e.to_string(),
+                "Inserting query must have the same schema length as the table."
+            );
+        }
+    }
```
I think you can write this much more concisely using [`unwrap_err`](https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_err)

```suggestion
    let e = write_df
        .write_table("t", DataFrameWriteOptions::new())
        .await
        .unwrap_err();
```

alamb replied:
The same comment applies to the code below as well

zhuqi-lucas replied:
Good idea, thanks @alamb , addressed in latest PR.

alamb commented on datafusion/core/tests/dataframe/mod.rs:
```diff
@@ -5274,3 +5274,61 @@ async fn register_non_parquet_file() {
+    let initial_table = Arc::new(MemTable::try_new(schema.clone(), vec![vec![]])?);
+    session_ctx.register_table("t", initial_table.clone())?;
+
+    // There are three cases we need to check
```
```suggestion
    // There are two cases we need to check
```

zhuqi-lucas replied:
Good catch! Addressed in latest

alamb reviewed (approved):
This looks like an improvment to me -- thank you @zhuqi-lucas and @jonahgao

zhuqi-lucas commented:
> This looks like an improvment to me -- thank you @zhuqi-lucas and @jonahgao

Thanks @alamb for review, addressed in latest PR. For union schema potential schema checking improvement https://github.com/apache/datafusion/pull/14572#discussion_r1954710305 , we may can create a new issue to discuss it.

alamb commented:
🚀
